### ISSUE TITLE
Exposing latest netcdf 4.9.x library functionality: quantize, zstandard

### ISSUE THREAD

**Issue by durack1** [PCMDI/cmor#721]
The latest versions of [libnetcdf](https://anaconda.org/conda-forge/libnetcdf) include new functions to further squash data using lossy compression, see [Charlie Zender: Why & How to Increase Dataset Compression in CMIP7](https://docs.google.com/presentation/d/1Ca26hD3KtSSy5tkWbVWHj261j2wJ2xjB_VHzQaFkpb0/edit#slide=id.gf0cc759aa2_0_177) - in particular the quantize and zstandard operations.

How easy is it to expose this in CMOR 3.8.1?

ping @taylor13 @matthew-mizielinski @sashakames @piotrflorek @czender

**Comment by durack1** [comment]
In the presentation (noted above), E3SM atmospheric data contributed to CMIP6 had a compression ratio of 1.8, so around half the lossless uncompressed size (netcdf deflate) vs vanilla writes.

To gauge the impact of using lossy compression on data usability, a couple of example cases would be great to test:
1. the impact of lossy compression (BitRound, BitGroom, GranularBR) when multiple layered operations occur (e.g. ocean data is regridded, then multiple separate fields (thetao, so, ..) are used to calculate a derived quantity - e.g. ocean heat content - so very preliminary discussion is contained in [Griffies et al., 2016 Appendix A](https://gmd.copernicus.org/articles/9/3231/2016/gmd-9-3231-2016.pdf#page=9) 
2. the impact of lossy compression on radiation calculations e.g. degraded cloud area fraction being able to recover correct radiation properties (e.g. @martinjuckes)

ping @juliettelavoie @geo-rao

**Comment by durack1** [comment]
@taylor13 agreed, it will be useful to catch these comments and redirect toward another place so that testing impacts on data usability and access can be undertaken.

This particular issue can be closed when we've ascertained how to expose these new netcdf functions through CMOR, and indeed whether this is available in the CMOR3 or CMOR4 (future) releases. If it is possible, having these available in a soon to be released version would be my preference, if this is relatively little work

**Comment by mauzey1** [comment]
It should be easy to expose [nc_def_var_quantize](https://docs.unidata.ucar.edu/netcdf-c/4.9.2/group__variables.html#ga669e4b23b9b7fb321b995982972cdd93) and [nc_def_var_zstandard](https://docs.unidata.ucar.edu/netcdf-c/4.9.2/netcdf__filter_8h.html#a01c702426610d3cd335da1430dfc9326) in the same way we do with [nc_def_var_deflate](https://docs.unidata.ucar.edu/netcdf-c/current/group__variables.html#ga59dad3301f241a7eb86f31b339af2d26) and [cmor_set_deflate](https://cmor.llnl.gov/mydoc_cmor3_api/#cmor_set_deflate).

We might need to add check for the version of NetCDF4 being used to determine if the functions are supported similar to the following code.
https://github.com/PCMDI/cmor/blob/047fd2c4b80769f3c2800f1dcbd0f930d36a51b1/Src/cmor.c#L25-L42

**Comment by matthew-mizielinski** [comment]
Another piece of context here; [Baker et al](https://gmd.copernicus.org/articles/9/4381/2016/gmd-9-4381-2016.pdf) added some data for lossy compression to an ensemble of CESM simulations. Several of the figures here highlight the benefits of and issues created by introducing lossy compression

**Comment by durack1** [comment]
> Another piece of context here; [Baker et al](https://gmd.copernicus.org/articles/9/4381/2016/gmd-9-4381-2016.pdf) added some data with lossy compression to an ensemble of CESM simulations. Several of the figures here highlight the benefits of and issues created by introducing it

Nice catch! It looks like depending on what you're calculating (Figures 5, 6, 9, 13) it does matter. Their section 6 "Lessons learned" is certainly worth reading, and notes that relationships between variables matter, focused on their surface energy balance anomaly using 4 separate variables - pointing out that commonly derived variables need to be a primary consideration, as do high frequency/precipitation extremes and other very data sensitive analyses. They also point out how missing/fill values are treated needs to be a consideration. It could be useful to loop around with [Allison](https://staff.ucar.edu/users/abaker) and Gary (@strandwg) to see if follow on analyses are available

**Comment by taylor13** [comment]
Thanks for the link to Baker.  Looks like useful information.

**Comment by sashakames** [comment]
Hey, a consideration to stick with "Lossy" compression and skip the "lossless" compression is that the latter is done in hardware on the storage layer and much more efficient that way, you are saving CPU cycles that way.  Just another consideration, and I didn't have a chance to raise that with Charlie yesterday.

**Comment by durack1** [comment]
> the latter is done in hardware on the storage layer and much more efficient that way, you are saving CPU cycles that way.

Is this on the GPFS hardware that you're talking about? So this is infrastructure/hardware dependent, right?

**Comment by sashakames** [comment]
typically the blockstorage controller.  All the US datacenters have it for sure, its a fairly standard feature been around for a while, but true not a guarantee that every center will enable that. (But I would be surprised if they don't by now)

**Comment by sashakames** [comment]
OTOH, the end-user will end up with a larger file size if downloaded locally, so there would be a trade off (but the downloading tool could then rewrite the file with L3 lossless compression to save space locally)

**Comment by czender** [comment]
That's news to me. What type of lossless compression do these blockstorage controllers implement? Zstandard? DEFLATE? Or...?
